### DISCUSSION TITLE
Software rendering on Linux Drm/FBDev backends

### DISCUSSION THREAD

**Discussion by MiKom** [AvaloniaUI/Avalonia#13710] [Q&A] [answered]
Hello.

I'm evaluating Avalonia for usage on an embedded device running Linux and I'm using the Drm backend.
I've noticed that on startup, the sample application takes almost 130MB of RAM.

I then checked my application on X11 desktop and the memory consumption was similar.

However, after I forced the software rendering on X11 (using `.With(new X11PlatformOptions { RenderingMode = new[] { X11RenderingMode.Software } })` on AppBuilder) I noticed that the RAM usage dropped to ~85MB. These additional 45MB of free memory may determine whether Avalonia will be viable on my target platform. I can live with less performant rendering as the application doesn't contain too much animations anyway.

Unfortunately, It seems I can't force software rendering when using Drm or FBDev backends. The `LinuxFramebufferPlatformOptions` only contains settings for FPS.

Is it possible to force software rendering on these embedded backends?

I may also not fully understand these backends and they somehow imply gpu rendering.

**Comment by thevortexcloud** [comment]
Software rendering with DRM does not make sense. DRM is the Direct Rendering Manager. The whole point of it is to provide an interface for talking to GPUs for rendering things.

https://en.wikipedia.org/wiki/Direct_Rendering_Manager

Wikipedia also has this diagram showing it.

https://upload.wikimedia.org/wikipedia/commons/6/6d/Access_to_video_card_with_DRM.svg

FBDev is the framebuffer device. Which also implies some sort of GPU. As the framebuffer refers to the framebuffer stored in VRAM on a GPU. Essentially, a GPU will output anything onto the screen put into it's framebuffer and this is how really everything is displayed on a screen. It's not typically hardware accelerated though when accessed directly. But it can be via things like DRI.

https://en.wikipedia.org/wiki/Linux_framebuffer
https://en.wikipedia.org/wiki/Framebuffer
https://en.wikipedia.org/wiki/Direct_Rendering_Infrastructure


That said, I am not sure how Avalonia does it's rendering with all the Linux backends. Assuming it's via OpenGL you could in theory force it by using a CPU OpenGL driver.

**Reply by kekekeks** [reply]
Technically it's possible to allocate a so-called dumb buffers in CPU memory and display those for displaying stuff on outputs. That's one of the few things that are guaranteed to be supported by drm system, IIRC

**Reply by MiKom** [reply]
Thanks @thevortexcloud and @kekekeks for all the information. I think I have what I need so far.

**Comment by kekekeks** [comment, marked as answer]
FBDev backend is always using software rasterizer, DRM one is always hw-accelerated with the current implementation

**Reply by MiKom** [reply]
Hey. Thanks for the info! I'll check for the memory consumption on both of these.

**Comment by DeanZhuo** [comment]
Got the same issue, memory behavior actively degrades over time:
- CmaFree (of 256MB CmaTotal) trends to 0 and stays there after 20-40 min of sustained rendering 
- Kernel Slab grows from ~44MB to 150-240MB+ over the same period
- System swap usage climbs steadily, eventually causing the UI to become fully unresponsive (libinput reports event processing lag)
- dmesg shows periodic [drm] Resetting GPU. messages roughly every ~30 min prior to full exhaustion

I ran the identical app via StartLinuxFbDev for almost 2 hours: CmaFree stayed healthy, Slab stayed flat (~45-46MB), swap never grew, and the app never froze.

Environment:
- Avalonia 12.1.0/12.1.1, Avalonia.LinuxFramebuffer
- Raspberry Pi 3B+, Raspberry Pi OS, vc4-drm driver (vc4-kms-v3d)
- App published with PublishAot=true, linux-arm64
- Running via StartLinuxDrm(args, card: null, options: new DrmOutputOptions { Scaling = 1.0 })
